File: ios/USTAQuery/ViewModels/RankingsViewModel.swift
import Foundation

@Observable
final class RankingsViewModel {
    static let shared = RankingsViewModel()

    private enum Keys {
        static let listKey = "ustaquery_rankings_listKey"
        static let gender = "ustaquery_rankings_gender"
        static let age = "ustaquery_rankings_age"
        static let section = "ustaquery_rankings_section"
    }

    var listKey: String {
        didSet { UserDefaults.standard.set(listKey, forKey: Keys.listKey) }
    }
    var gender: String {
        didSet { UserDefaults.standard.set(gender, forKey: Keys.gender) }
    }
    var ageRestriction: String {
        didSet { UserDefaults.standard.set(ageRestriction, forKey: Keys.age) }
    }
    var page = 0
    var publishDate = ""

    var data: PagedResponse<Ranking>?
    var versions: [String] = []
    var sections: [String] = []
    var isLoading = false
    var error: String?

    var sectionFilter: String {
        didSet {
            UserDefaults.standard.set(sectionFilter.isEmpty ? nil : sectionFilter, forKey: Keys.section)
        }
    }

    var selectedList: (value: String, label: String, catalogPattern: String) {
        AppConstants.listTypes.first(where: { $0.value == listKey }) ?? AppConstants.listTypes[0]
    }

    var catalogId: String {
        AppConstants.buildCatalogId(pattern: selectedList.catalogPattern, gender: gender, age: ageRestriction)
    }

    var genderLabel: String {
        AppConstants.genders.first(where: { $0.value == gender })?.label ?? gender
    }

    var ageLabel: String {
        AppConstants.ageGroupLabels[ageRestriction] ?? ageRestriction
    }

    private var lastVersionsCatalogId = ""

    init() {
        let defaults = UserDefaults.standard
        self.listKey = defaults.string(forKey: Keys.listKey) ?? "STANDING"
        self.gender = defaults.string(forKey: Keys.gender) ?? "M"
        self.ageRestriction = defaults.string(forKey: Keys.age) ?? "Y12"
        self.sectionFilter = defaults.string(forKey: Keys.section) ?? ""
    }

    func loadIfNeeded() async {
        guard data == nil else { return }
        await fetchAll()
    }

    func reload() async {
        lastVersionsCatalogId = ""
        await fetchAll()
    }

    func updateFilter(list: String? = nil, gender g: String? = nil, age: String? = nil, section: String? = nil) {
        let catalogChanging = list != nil || g != nil || age != nil
        if let list { listKey = list }
        if let g { gender = g }
        if let age { ageRestriction = age }
        if let section { sectionFilter = section }
        if catalogChanging {
            page = 0
            publishDate = ""
        }
        Task { await fetchAll() }
    }

    func updateVersion(_ date: String) {
        publishDate = date
        page = 0
        Task { await fetchLeaderboard() }
    }

    func goToPage(_ p: Int) {
        page = p
        Task { await fetchLeaderboard() }
    }

    private func fetchAll() async {
        let currentCatalogId = catalogId
        if currentCatalogId != lastVersionsCatalogId {
            lastVersionsCatalogId = currentCatalogId
            await fetchVersions(currentCatalogId)
            await fetchSections(currentCatalogId)
        }
        await fetchLeaderboard()
    }

    private func fetchVersions(_ catId: String) async {
        do {
            versions = try await APIClient.getRankingVersions(catalogId: catId)
            if !versions.isEmpty && publishDate.isEmpty {
                publishDate = versions[0]
            }
        } catch {
            versions = []
        }
    }

    private func fetchSections(_ catId: String) async {
        do {
            sections = try await APIClient.getRankingSections(catalogId: catId)
        } catch {
            sections = []
        }
    }

    private func fetchLeaderboard() async {
        isLoading = true
        error = nil
        do {
            data = try await APIClient.getLeaderboard(
                catalogId: catalogId,
                page: page,
                size: 50,
                publishDate: publishDate.isEmpty ? nil : publishDate,
                section: sectionFilter.isEmpty ? nil : sectionFilter
            )
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
}

File: android/app/src/main/java/com/usta/query/viewmodel/RankingsViewModel.kt
package com.usta.query.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.usta.query.data.api.ApiClient
import com.usta.query.data.model.Ranking
import kotlinx.coroutines.launch

class RankingsViewModel : ViewModel() {
    var listKey by mutableStateOf("STANDING")
    var gender by mutableStateOf("M")
    var ageRestriction by mutableStateOf("Y12")
    var page by mutableIntStateOf(0)
    var publishDate by mutableStateOf("")

    var data by mutableStateOf<List<Ranking>?>(null)
    var totalPages by mutableIntStateOf(0)
    var versions by mutableStateOf<List<String>>(emptyList())
    var isLoading by mutableStateOf(false)
    var error by mutableStateOf<String?>(null)

    val catalogId: String
        get() = buildCatalogId(listKey, gender, ageRestriction)

    fun loadIfNeeded() {
        if (data == null) fetchAll()
    }

    fun reload() {
        page = 0
        fetchAll()
    }

    fun updateFilter(list: String? = null, g: String? = null, age: String? = null) {
        list?.let { listKey = it }
        g?.let { gender = it }
        age?.let { ageRestriction = it }
        page = 0
        publishDate = ""
        fetchAll()
    }

    fun updateVersion(date: String) {
        publishDate = date
        page = 0
        fetchLeaderboard()
    }

    fun goToPage(p: Int) {
        page = p
        fetchLeaderboard()
    }

    private fun fetchAll() {
        viewModelScope.launch {
            try {
                versions = ApiClient.api.getRankingVersions(catalogId)
                if (versions.isNotEmpty() && publishDate.isEmpty()) {
                    publishDate = versions[0]
                }
            } catch (_: Exception) {
                versions = emptyList()
            }
            fetchLeaderboard()
        }
    }

    private fun fetchLeaderboard() {
        isLoading = true
        viewModelScope.launch {
            try {
                val resp = ApiClient.api.getLeaderboard(
                    catalogId = catalogId,
                    page = page,
                    size = 50,
                    publishDate = publishDate.takeIf { it.isNotEmpty() }
                )
                data = resp.content
                totalPages = resp.totalPages
            } catch (e: Exception) {
                error = e.message
            } finally {
                isLoading = false
            }
        }
    }

    companion object {
        fun buildCatalogId(listKey: String, gender: String, age: String): String {
            val pattern = when (listKey) {
                "STANDING" -> "JUNIOR_NULL_{G}_STANDING_{A}_UNDER_NULL_NULL_NULL"
                "SEEDING_SINGLES" -> "JUNIOR_NULL_{G}_SEEDING_{A}_UNDER_SINGLES_NULL_NULL"
                "SEEDING_DOUBLES" -> "JUNIOR_NULL_{G}_SEEDING_{A}_UNDER_DOUBLES_INDIVIDUAL_NULL"
                "BONUS_POINTS" -> "JUNIOR_NULL_{G}_BONUS_POINTS_{A}_UNDER_NULL_NULL_NULL"
                "QUOTA" -> "JUNIOR_NULL_{G}_QUOTA_{A}_UNDER_NULL_NULL_S05"
                "YEAR_END_COMBINED" -> "JUNIOR_NULL_{G}_YEAR_END_{A}_UNDER_NULL_NULL_NULL"
                "YEAR_END_DOUBLES" -> "JUNIOR_NULL_{G}_YEAR_END_{A}_UNDER_DOUBLES_INDIVIDUAL_NULL"
                else -> "JUNIOR_NULL_{G}_STANDING_{A}_UNDER_NULL_NULL_NULL"
            }
            return pattern.replace("{G}", gender).replace("{A}", age)
        }

        val listTypes = listOf(
            Triple("STANDING", "Combined National Standing List", ""),
            Triple("SEEDING_SINGLES", "Singles Seeding List", ""),
            Triple("SEEDING_DOUBLES", "Doubles Seeding List", ""),
            Triple("BONUS_POINTS", "Bonus Points List", ""),
            Triple("QUOTA", "Quota List", ""),
            Triple("YEAR_END_COMBINED", "Final Year End Combined Rank List", ""),
            Triple("YEAR_END_DOUBLES", "Final Year End Doubles Rank List", "")
        )

        val genders = listOf("M" to "Boys", "F" to "Girls")
        val ageRestrictions = listOf("Y12", "Y14", "Y16", "Y18")
        val ageGroupLabels = mapOf("Y12" to "12 & Under", "Y14" to "14 & Under", "Y16" to "16 & Under", "Y18" to "18 & Under")
    }
}
